core/module.cc

```diff
@@ -245,57 +245,77 @@ void Module::DeregisterAllAttributes() {
   }
 }
 
-placement_constraint Module::ComputePlacementConstraints() const {
+placement_constraint Module::ComputePlacementConstraints(
+    std::unordered_set<const Module *> &visited) const {
   // Take the constraints we have.
   int constraint = node_constraints_;
-  for (size_t i = 0; i < ogates_.size(); i++) {
-    if (ogates_[i]) {
-      auto next = static_cast<Module *>(ogates_[i]->arg());
-      // Restrict constraints to account for other modules in the pipeline.
-      constraint &= next->ComputePlacementConstraints();
-      if (constraint == 0) {
-        LOG(WARNING) << "At " << name_
-                     << " after accounting for constraints from module "
-                     << next->name_ << " no feasible placement exists.";
+  // Follow the rest of the tree unless we have already been here.
+  if (visited.find(this) == visited.end()) {
+    visited.insert(this);
+    for (size_t i = 0; i < ogates_.size(); i++) {
+      if (ogates_[i]) {
+        auto next = static_cast<Module *>(ogates_[i]->arg());
+        // Restrict constraints to account for other modules in the pipeline.
+        constraint &= next->ComputePlacementConstraints(visited);
+        if (constraint == 0) {
+          LOG(WARNING) << "At " << name_
+                       << " after accounting for constraints from module "
+                       << next->name_ << " no feasible placement exists.";
+        }
       }
     }
   }
   return constraint;
 }
 
-void Module::AddActiveWorker(int wid) {
-  active_workers_.insert(wid);
-  if (propagate_workers_) {
-    for (auto ogate : ogates_) {
-      if (ogate) {
-        auto next = static_cast<Module *>(ogate->arg());
-        next->AddActiveWorker(wid);
+void Module::AddActiveWorker(int wid, ModuleTask *t) {
+  uint64_t worker_mask = 1ull << wid;
```

Comment:
No big reason.  I guess in the back of my mind, if we were trying to be very general, then we would allow for an arbitrary number of workers, in which case we would benefit from a set for which we don't need to know the max value in advance.  But if we assume we know how many workers are possible (e.g. max workers) then vector<bool> is good.